optimize.sh

```diff
@@ -4,37 +4,43 @@ command -v shellcheck > /dev/null && shellcheck "$0"
 
 export PATH=$PATH:/root/.cargo/bin
 
-mkdir -p artifacts
-contractdir="$1"
-
-# There are two cases here
-# 1. All contracts (or one) are included in the root workspace  (eg. `cosmwasm-template`, `cosmwasm-examples`, `cosmwasm-plus`)
-#    In this case, we pass no argument, just mount the proper directory.
-# 2. Contacts are excluded from the root workspace, but import relative paths from other packages (only `cosmwasm`).
-#    In this case, we mount root workspace and pass in a path `docker run <repo> ./contracts/hackatom`
-
-# This parameter allows us to mount a folder into docker container's "/code"
-# and build "/code/contracts/mycontract".
-# Note: if contractdir is "." (default in Docker), this ends up as a noop
-echo "Building contract in $(realpath -m "$contractdir")"
-(
-  cd "$contractdir"
-
-  # Linker flag "-s" for stripping (https://github.com/rust-lang/cargo/issues/3483#issuecomment-431209957)
-  # Note that shortcuts from .cargo/config are not available in source code packages from crates.io
-  RUSTFLAGS='-C link-arg=-s' cargo build --release --target wasm32-unknown-unknown --locked
-)
-
-# wasm-optimize on all results
-for wasm in "$contractdir"/target/wasm32-unknown-unknown/release/*.wasm; do
-  name=$(basename "$wasm")
-  echo "Optimizing $name"
-  wasm-opt -Os "$wasm" -o "artifacts/$name"
+contracts_root="$1"
+
+# See https://stackoverflow.com/a/23357277/2013738 to understand this find-to-array monster
+contract_dirs=()
+while IFS=  read -r -d $'\0'; do
+  contract_dir=$(dirname "$REPLY")
+  contract_dirs+=("$contract_dir")
+done < <(find "$contracts_root" \( -name target -o -name package -o -name .git \) -type d -prune -false -o -name Cargo.toml -type f -print0)
+echo "Found the following contract dirs: ${contract_dirs[*]}"
+
+for contract_dir in "${contract_dirs[@]}"; do
+  echo "Building contract in $(realpath -m "$contract_dir")"
+  (
+    cd "$contract_dir"
```

Comment:
The issue of the `.` directory being called in any random order and overriding others can be fixed here if we just filter for actual crates (eg. if there is the `members` field in Cargo.toml skip it)

Otherwise, I like the find